modules/core/src/main/java/org/apache/ignite/cache/validation/IgnitePluggableSegmentationResolver.java

```diff
@@ -0,0 +1,275 @@
+
+        ctx.discovery().setCustomEventListener(
+            ChangeGlobalStateFinishMessage.class,
+            new ClusterStateChangedEventListener()
+        );
+
+        ctx.internalSubscriptionProcessor().registerDistributedConfigurationListener(
+            new DistributedConfigurationLifecycleListener() {
+                /** {@inheritDoc} */
+                @Override public void onReadyToRegister(DistributedPropertyDispatcher dispatcher) {
+                    dispatcher.registerProperty(segResolverEnabledProp);
+                }
+
+                /** {@inheritDoc} */
+                @Override public void onReadyToWrite() {
+                    setDefaultValue(segResolverEnabledProp, U.isLocalNodeCoordinator(ctx.discovery()), log);
+                }
+            });
+    }
+
+    /** @return Discovery data. */
+    public Serializable provideDiscoveryData() {
+        return state;
+    }
+
+    /** @param data Discovery data. */
+    public void onDiscoveryDataReceived(Serializable data) {
+        state = (State)data;
+    }
+
+    /** @param node Node. */
+    public void validateNewNode(ClusterNode node) {
+        if (node.isClient())
+            return;
+
+        if (!TRUE.equals(node.attribute(ATTR_SEG_RESOLVER_CONFIGURED))) {
+            throw new IgniteException( "The Segmentation Resolver plugin is not configured for the server node that is" +
+                " trying to join the cluster. Since the Segmentation Resolver is only applicable if all server nodes" +
+                " in the cluster have one, node join request will be rejected [rejectedNodeId=" + node.id() + ']');
+        }
+
+        if (state == State.VALID) {
+            DiscoCache discoCache = ctx.discovery().discoCache(new AffinityTopologyVersion(lastCheckedTopVer, 0));
+
+            if (discoCache != null) {
+                for (ClusterNode srv : discoCache.serverNodes()) {
+                    if (!ctx.discovery().alive(srv))
+                        throw new IgniteException("Node join request will be rejected due to concurrent node left" +
+                            " process handling [rejectedNodeId=" + node.id() + ']');
+                }
+            }
+        }
+    }
+
+    /** */
+    private boolean isDisabled() {
+        Boolean res = segResolverEnabledProp.get();
+
+        return res == null || !res;
+    }
+
+    /** @return return. */
+    private String formatTopologyNodes(Collection<ClusterNode> nodes) {
+        return nodes.stream().map(n -> n.id().toString()).collect(Collectors.joining(", "));
+    }
+
+    /** */
+    private class TopologyChangedEventListener implements DiscoveryEventListener, HighPriorityListener {
+        /** {@inheritDoc} */
+        @Override public void onEvent(DiscoveryEvent evt, DiscoCache discoCache) {
+            lastCheckedTopVer = evt.topologyVersion();
+
+            if (isDisabled())
+                return;
+
+            if (state == State.VALID && evt.type() == EVT_NODE_FAILED) {
```

Comment:
When we change cluster state **to** read-only is treated as CLUSTER_WRITE_BLOCKED, but if we start the cluster in the read-only mode (`clusterStateOnStart` config property) it will be treated as "VALID" and this code block executes on segmentation.
MB we should check the cluster state also.

Reply:
Done.